src/blk/kernel/KernelDevice.cc

```diff
@@ -719,7 +722,11 @@ void KernelDevice::_discard_thread()
 
 int KernelDevice::queue_discard(interval_set<uint64_t> &to_release)
 {
-  if (!support_discard)
+  // device not for async discard
+  if (!support_discard || !cct->_conf->bdev_enable_discard || !cct->_conf->bdev_async_discard)
```

Comment:
the same here - following check for discard_thread.is_started() should be sufficient